biz.aQute.resolve/src/biz/aQute/resolve/Bndrun.java

```diff
@@ -0,0 +1,146 @@
+			for (Resource resource : resources) {
+				VersionedClause runBundle = ResourceUtils.toVersionClause(resource, "[===,==+)");
+				if (!runBundles.contains(runBundle)) {
+					runBundles.add(runBundle);
+				}
+			}
+			Collections.sort(runBundles, new Comparator<VersionedClause>() {
+				@Override
+				public int compare(VersionedClause a, VersionedClause b) {
+					int diff = a.getName().compareTo(b.getName());
+					return (diff != 0) ? diff : a.getVersionRange().compareTo(b.getVersionRange());
+				}
+			});
+
+			File runFile = getPropertiesFile();
+			BndEditModel bem = new BndEditModel(getWorkspace());
+			Document doc = new Document(IO.collect(runFile));
+			bem.loadFrom(doc);
+
+			List<VersionedClause> bemRunBundles = bem.getRunBundles() != null ? bem.getRunBundles()
+					: new ArrayList<VersionedClause>();
+
+			String originalRunbundlesString = headerClauseListFormatter.convert(bemRunBundles);
+			logger.debug("Original -runbundles was:\n\t {}", originalRunbundlesString);
+			String runbundlesString = headerClauseListFormatter.convert(runBundles);
+			logger.debug("Resolved -runbundles is:\n\t {}", runbundlesString);
+
+			List<VersionedClause> deltaAdd = new ArrayList<>(runBundles);
+			deltaAdd.removeAll(bemRunBundles);
+			List<VersionedClause> deltaRemove = new ArrayList<>(bemRunBundles);
+			deltaRemove.removeAll(runBundles);
+			boolean added = bemRunBundles.addAll(deltaAdd);
+			boolean removed = bemRunBundles.removeAll(deltaRemove);
+			if (added || removed) {
+				if (failOnChanges && !bemRunBundles.isEmpty()) {
+					String message = String.format("The runbundles have changed. Failing the build!\nWas: %s\nIs: %s",
+							originalRunbundlesString, runbundlesString);
+					throw new Exception(message);
+				}
+				bem.setRunBundles(bemRunBundles);
+				logger.info("Writing changes to {}", runFile.getAbsolutePath());
+				logger.info("{}:{}", Constants.RUNBUNDLES, bem.getDocumentChanges().get(Constants.RUNBUNDLES));
+				bem.saveChangesTo(doc);
+				if (writeOnChanges) {
+					IO.store(doc.get(), runFile);
+				}
+			}
+			return headerClauseListFormatter.convert(bem.getRunBundles());
+		}
+	}
+
+	public void setRunBundles(Collection< ? extends Container> runBundles) throws Exception {
+		Collection<Container> runbundles = getRunbundles();
+		runbundles.clear();
+		runbundles.addAll(runBundles);
+	}
+
+	public void test(File targetDir) throws Exception {
+		logger.info("Test " + this);
+		ProjectTester projectTester = getProjectTester();
+		File dir = new File(targetDir, "tmp");
+		IO.delete(dir);
+		dir.mkdirs();
+
+		projectTester.setCwd(dir);
+		projectTester.setReportDir(targetDir);
+		projectTester.test();
+
+		if (!getWarnings().isEmpty()) {
+			logger.info("Warnings\n{}", Strings.join("\n", getWarnings()));
+		}
+
+		IO.delete(dir);
+
+		if (getErrors().isEmpty())
+			return;
+
+		logger.info("Errors\n{}", Strings.join("\n", getErrors()));
+		throw new Exception("Test had errors");
```

Comment:
Also don't throw exception here. The errors should be in the Bndrun which the caller should check.